### PR TITLE
netty: switch plaintext upgrade negotiator to new format

### PR DESCRIPTION
**Pull request by carl-mastrangelo**

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
The test is broken ATM, it just passes without actually doing negotiation.  Either the server upgrade handler is broken in netty, or I don't know what server upgrade looks like and my expectation is wrong.   Since there are basically no other tests for it, I'm not sure which is which.

**Comment by carl-mastrangelo** [comment]
@ejona86 this is ready for review.

(the reason the test failed is that both the http client codec AND the upgrade codec need to be in the pipeline, not just the upgrade.  LocalChannel just takes the writes it gets and gives them to the LocalServerChannel, skipping any check that the msg is a ByteBuf.   This is why the request showed up as valid on the server; it was never encoded and thus skips the normal HTTP processing.  Even if it did throw, the initial write doesn't listen for errors, so the program just silently hangs.  Both of these problems have been fixed in this version.)

**Comment by ejona86** [comment]
> the reason the test failed is that both the http client codec AND the upgrade codec need to be in the pipeline, not just the upgrade

Hmmm... I had noticed that, but when I added it locally the test still failed. Anyway, glad you figured it out.

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java [hunk 1]
```diff
@@ -407,31 +407,79 @@ public HostPort(String host, int port) {
...
+      HttpClientUpgradeHandler upgrader =
+          new HttpClientUpgradeHandler(httpClientCodec, upgradeCodec, /*maxContentLength=*/ 1000);
+      ctx.pipeline().addBefore(ctx.name(), null, upgrader);
+
+      // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
+      // which causes the upgrade headers to be added
+      DefaultHttpRequest upgradeTrigger =
+          new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+      upgradeTrigger.headers().add(HttpHeaderNames.HOST, authority);
+      ctx.writeAndFlush(upgradeTrigger).addListener(FireExceptionListener.INSTANCE);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java]: ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java [hunk 2]
```diff
@@ -690,31 +673,46 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
+       */
+      Http2ClientUpgradeCodec upgradeCodec = new Http2ClientUpgradeCodec(next);
+      HttpClientCodec httpClientCodec = new HttpClientCodec();
+      HttpClientUpgradeHandler upgrader =
+          new HttpClientUpgradeHandler(httpClientCodec, upgradeCodec, /*maxContentLength=*/ 1000);
+      ctx.pipeline().addBefore(ctx.name(), null, upgrader);
+
       // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
       // which causes the upgrade headers to be added
       DefaultHttpRequest upgradeTrigger =
```
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java]: Arg. Disregard. This is from an earlier partial review.

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java [hunk 3]
```diff
@@ -727,39 +725,46 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
...
-
   /**
    * Adapts a {@link ProtocolNegotiationEvent} to the {@link GrpcHttp2ConnectionHandler}.
    */
   static final class GrpcNegotiationHandler extends ChannelInboundHandlerAdapter {
-    private final GrpcHttp2ConnectionHandler next;
+    private final GrpcHttp2ConnectionHandler grpcHttp2ConnectionHandler;
+    private final ChannelHandler next;
+
+    public GrpcNegotiationHandler(
```
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java]: Arg. Disregard. This is from an earlier partial review.

File: netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java (hunk 1 above)
Comment (carl-mastrangelo) [reply netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java]: TIL.  

I remembered the CLOSE one, but didn't realize the exception one.

**Pull request merged**